### PR TITLE
Provide a Miou implementation of dns-client

### PR DESCRIPTION
**Pull request by dinosaure**
A free implementation of `dns-client` with Miou (and `happy-eyeballs-miou-unix`).

### CODE REVIEW

File: dns-client-miou-unix.opam [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  "happy-eyeballs-miou-unix"
+]
+synopsis: "DNS client API for Miou"
+description: """
+A client implementation using uDNS using Miou.
+"""
+
+pin-depends: [
+  [ "happy-eyeballs-miou-unix.1.1.0" "git+https://github.com/robur-coop/happy-eyeballs.git#2a2f276b547d9ac34c0ce9ca40a19b67c7db726f" ]
+]
```
Comment (hannesm) [inline dns-client-miou-unix.opam:33]:
```suggestion
```

**Comment by hannesm** [comment]
from my perspective, this is ready to go!? I just merged the happy-eyeballs to opam-repository and removed the pin-depend here.

File: dns-client-miou-unix.opam [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+build: [
+  [ "dune" "subst"] {dev}
+  [ "dune" "build" "-p" name "-j" jobs ]
+  [ "dune" "runtest" "-p" name "-j" jobs ] {with-test}
+]
+
+depends: [
+  "dune" {>="2.0.0"}
+  "ocaml" {>= "4.08.0"}
```
Comment (hannesm) [inline dns-client-miou-unix.opam:17]:
```suggestion
  "ocaml" {>= "5.0.0"}
```

File: miou/client/dns_client_miou_unix.ml [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            let rx_len' = rx_len - (expected_len + 2) in
+            Bytes.blit buf (expected_len + 2) buf' 0 rx_len';
+            go buf' rx_len'
+        | Some expected_len when Bytes.length buf >= expected_len + 2 ->
+          let len = (expected_len + 2) - rx_len in
+          Tls_miou_unix.really_read fd buf ~off:rx_len ~len;
+          go buf (rx_len + len)
+        | Some expected_len ->
+          let buf' = Bytes.make (expected_len + 2) '\000' in
+          Bytes.blit buf 0 buf' 0 rx_len;
```
Comment (hannesm) [inline miou/client/dns_client_miou_unix.ml:120]: I'm actually unsure about `buf'` -- what is it's purpose? we allocate and blit into it, but then we throw it away!?

Comment (hannesm) [reply miou/client/dns_client_miou_unix.ml:120]: I suspect the line below should use `buf'`. The `go` function is a bit hard to read and understand imho...

Comment (dinosaure) [reply miou/client/dns_client_miou_unix.ml:120]: If I read correctly, this is the worst case where the current `buf` is not large enough to contains the actual DNS packet. In this branch, the idea is to allocate a new bytes `buf'`) which can fit the DNS packet and use it (and forget the old `buf`). Actually, we have an error on the `go buf rx_len`, it should be `go buf' rx_len`.

Comment (dinosaure) [reply miou/client/dns_client_miou_unix.ml:123]: Fixed, I added a comment to explain what is the purpose of this branch.

**Pull request merged**